### PR TITLE
Chapter7/code coverage

### PR DESCRIPTION
**Pull request by bernedom**
Add examples to produce a code coverage report and for building with various sanitizers enabled

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -99,3 +101,30 @@ jobs:
       - name: pytest
         run: python3 -m pytest -x -v
 
+      # Note in order for ctest -T Coverage to work the tests have to be executed first
+      # with ctest. Moving this before the test step breaks coverage
+      - name: coverage
+        if: contains(matrix.compiler, 'g++-10')
+        run: |
+          ctest -T test -T Coverage
```
Comment (mustafakemalgilor) [inline .github/workflows/ci.yml:109]: The "ctest -T test -T Coverage" invocation may be a separate step if artifacts are preserved between steps.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -99,3 +101,30 @@ jobs:
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: code-coverage-report
+          path: ./build/coverage-report.html
+
+
+      - name: sanitizers
+        if: contains(matrix.compiler, 'clang++-12')
+        run: |
+          cmake -S .. -B . -DCMAKE_BUILD_TYPE=Debug -DENABLE_ADDRESS_SANITIZER=ON -DENABLE_SANITIZER_ADDRESS=ON -DENABLE_SANITIZER_UNDEFINED_BEHAVIOR=ON
```
Comment (mustafakemalgilor) [inline .github/workflows/ci.yml:124]: I think we should avoid in-source builds, it may mess things up in the future.

File: CMakeLists.txt [hunk 3]
```diff
@@ -16,6 +16,7 @@ project(
         "The 'build-all' instructions for all examples for the book CMake Best Practices"
     LANGUAGES CXX
 )
+include(CTest)
 enable_testing()
```
Comment (mustafakemalgilor) [inline CMakeLists.txt:20]: include(CTest) makes enable_testing redundant,

Quoting from CMake documentation for enable_testing() (https://cmake.org/cmake/help/latest/command/enable_testing.html):

```
... This command is automatically invoked when the CTest module is included, except if the BUILD_TESTING option is turned off. ...
```

Comment (bernedom) [reply CMakeLists.txt:20]: True, will remove `enable_testing()`

File: .github/workflows/ci.yml (hunk 2 above)
Comment (bernedom) [reply .github/workflows/ci.yml:124]: This is not an in source build. I changed it to make it clearer.

**Pull request merged**